src/smtencoding/FStar.SMTEncoding.Encode.fs

```diff
@@ -1413,9 +1471,15 @@ let encode_free_var env fv tt t_norm quals =
          else let encode_non_total_function_typ = lid.nsstr <> "Prims" in
               let formals, (pre_opt, res_t) =
                 let args, comp = curried_arrow_formals_comp t_norm in
+                let comp =
+                  if is_reifiable_comp env.tcenv comp
+                  then S.mk_Total (TcUtil.reify_comp ({env.tcenv with lax=true}) (U.lcomp_of_comp comp) U_unknown)
```

Comment:
Right now, given `f: t -> St t'`, our approach is to just encode `f` at the type `t -> reify_comp (St t')`. 

This approach basically just removed effectful computations from the SMT encoding altogether, blurring the distinction between `f` and `reify f`.

Maybe we should instead retain a `reify_term : Term -> Term` and `reify_type : Term -> Term` in the SMT encoding.

We could have an SMT axiom saying `HasType f t ==> HasType (reify_term f) (reify_type t)`

And then, given `f : (t -> St t)`, we produce `HasType f (encode (t -> St t))` and `reify_type (encode (t -> St t)) = encode (t -> reify_comp (St t))`

This seems to me to be more regular and more faithful to the source typing discipline.